### ISSUE TITLE
Resolve and apply block metadata at start of block rule

### ISSUE THREAD

**Issue by mojavelinux** [opendevise/asciidoc-parsing-lab#26]
We've learned that block metadata lines, which are comprised of block attribute lines and the block title line, are subtly coupled with block parsing. In order to parse the block, there are instances when the block style is needed (e.g., a discrete heading). To resolve the block attribute values, the parser needs to know which block is coming in order to know what positional attributes are mapped, if any, so it knows which ones to treat as content attributes.

What we've done thus far is to wrap the value of every non-common block attribute in a deferred function to call once all the necessary information is known (this is "resolving the attribute value"). The question becomes, when are these values resolved? Currently, they are resolved in the action for the block. However, for compound blocks (i.e., blocks that accept child blocks), this comes too late. It would mean that the inline parsing for a content attribute would happen after the block and all its children are parsed, and hence out of document order.

A solution to this problem is to resolve the block attribute values as soon as the block is positively identified. Doing so will require the use of an action that runs on the rule before the contents of the block are matched. Here's how that might look:

```
sidebar = 
  metadata:(metadataKey:offset openingDelim:sidebar_delimiter_line &{ return enterBlock(context, openingDelim) } { return resolveBlockMetadata(metadataKey) })
  blocks:(lf* @(discrete_heading / listing / example / sidebar / list / indented / image / paragraph))*
  closingDelim:(lf* @(sidebar_delimiter_line / eof))
  {
    const delimiter = exitBlock(context)
    const node = { name: 'sidebar', type: 'block', form: 'delimited', delimiter, blocks, location: toSourceLocation(getLocation()) }
    return metadata ? Object.assign(node, { metadata }) : node
  }
```

Notice that it's necessary to pass the metadata through a global cache indexed by offset so the action can access it.

The benefit of this approach is that the resolved metadata will be available to the action block for the block rule, thus allowing the action to assign the metadata to the block at the time the node is created. 

An alternative strategy, which we aren't going to pursue at this time, would be to defer all inline parsing until after the entire AST of the document is built by the block grammar parser. In this scenario, the action for each block would define a resolve method on the node to defer and abstract away this work. The action for the document would then walk the tree and call this method on each node in succession. This would guarantee that all inline parsing would happen in document order. (It's also a more pure approach to parsing in general as it separates the lexer and semantic parser). However, there may be unforeseen side effects to this strategy that we aren't yet considering. More research is likely needed.

**Comment by mojavelinux** [comment]
There are actually two steps:

* process
* apply

The process step performs the following tasks:

* remaps positional attributes to named attributes, as specified by the posattrs argument
* stores options and roles as an array
* resolves attribute values by running the deferred value function, if applicable, prepares a map of promoted attributes

The apply step performs the following tasks:

* applies promoted attributes to the block
* assigns the metadata property to the block

**Comment by mojavelinux** [comment]
To reiterate, the idea is to process the block metadata before any child nodes (blocks or inlines) are parsed. This could happen as early as in the block rule as soon as the block is deterministically / unequivocally matched (in other words, the rule isn't going to fail).
